### PR TITLE
[AIDAPP-370]: Restrict the ability to grant overly permissive rights to users.

### PR DESCRIPTION
**Pull request by kandarp-canyon**
### Ticket(s) or GitHub Issue

- https://canyongbs.atlassian.net/browse/AIDAPP-370

### Technical Description

> Restrict the ability to grant overly permissive rights to users.

### Any deployment steps required?

> No.

### Are any Feature Flags and/or Data Migrations that can eventually be removed Added?

> Feature flag SuperAdminRole.
> 2024_12_04_175613_data_activate_super_admin_role_feature_flag.php.

_______________________________________________

### CODE REVIEW

**Review by Orrison — CHANGES_REQUESTED** [review]

File: app-modules/authorization/src/Models/Role.php [hunk 1]
```diff
@@ -80,7 +81,7 @@ public function scopeWeb(Builder $query): void
     public function scopeSuperAdmin(Builder $query): void
     {
         $query
-            ->where('name', 'authorization.super_admin')
+            ->where('name', Authenticatable::SUPER_ADMIN_ROLE)
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app-modules/authorization/src/Models/Role.php]: Done

File: app-modules/authorization/database/migrations/2024_12_02_151546_data_change_authorization.super_admin_to_sass_global_admin_in_roles_table.php
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app-modules/authorization/database/migrations/2024_12_02_151546_data_change_authorization.super_admin_to_sass_global_admin_in_roles_table.php:1]: Added feature flag.

File: app-modules/authorization/tests/Feature/Filament/Resources/User/EditUserTest.php [hunk 2]
```diff
@@ -147,3 +152,40 @@
     expect($second->isImpersonated())->toBeTrue();
     expect(auth()->id())->toBe($second->id);
 });
+
+it('Not allows user which has not SaaS Global Admin role to assign SaaS Global Admin role to other user', function () {
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app-modules/authorization/tests/Feature/Filament/Resources/User/EditUserTest.php]: updated

File: app-modules/contact/src/Filament/Resources/ContactResource/RelationManagers/ServiceRequestsRelationManager.php [hunk 3]
```diff
@@ -70,7 +71,7 @@ public function table(Table $table): Table
         return $table
             ->recordTitleAttribute('id')
             ->modifyQueryUsing(function ($query) {
-                $query->when(! auth()->user()->hasRole('authorization.super_admin'), function (Builder $q) {
+                $query->when(! auth()->user()->hasRole(Authenticatable::SUPER_ADMIN_ROLE), function (Builder $q) {
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app-modules/contact/src/Filament/Resources/ContactResource/RelationManagers/ServiceRequestsRelationManager.php]: Added.

File: app-modules/service-management/src/Filament/Resources/ServiceRequestResource/Pages/CreateServiceRequest.php [hunk 4]
```diff
@@ -84,7 +85,7 @@ public function form(Form $form): Form
                 Grid::make()
                     ->schema([
                         Select::make('type_id')
-                            ->options(ServiceRequestType::when(! auth()->user()->hasRole('authorization.super_admin'), function (Builder $query) {
+                            ->options(ServiceRequestType::when(! auth()->user()->hasRole(Authenticatable::SUPER_ADMIN_ROLE), function (Builder $query) {
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app-modules/service-management/src/Filament/Resources/ServiceRequestResource/Pages/CreateServiceRequest.php]: Added.

File: app-modules/service-management/src/Filament/Resources/ServiceRequestResource/Pages/EditServiceRequest.php [hunk 5]
```diff
@@ -100,7 +101,7 @@ public function form(Form $form): Form
                                         fn (ServiceRequest $record) => ServiceRequestType::withTrashed()
                                             ->whereKey($record->priority?->type_id)
                                             ->orWhereNull('deleted_at')
-                                            ->when(! auth()->user()->hasRole('authorization.super_admin'), function (Builder $query) {
+                                            ->when(! auth()->user()->hasRole(Authenticatable::SUPER_ADMIN_ROLE), function (Builder $query) {
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app-modules/service-management/src/Filament/Resources/ServiceRequestResource/Pages/EditServiceRequest.php]: Added.

File: app-modules/service-management/src/Filament/Resources/ServiceRequestResource/Pages/ListServiceRequests.php [hunk 6]
```diff
@@ -78,7 +79,7 @@ public function table(Table $table): Table
                 ],
                 'status',
             ])
-                ->when(! auth()->user()->hasRole('authorization.super_admin'), function (Builder $q) {
+                ->when(! auth()->user()->hasRole(Authenticatable::SUPER_ADMIN_ROLE), function (Builder $q) {
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app-modules/service-management/src/Filament/Resources/ServiceRequestResource/Pages/ListServiceRequests.php]: Added.

File: app-modules/service-management/src/Filament/Resources/ServiceRequestResource/Pages/ListServiceRequests.php [hunk 7]
```diff
@@ -164,7 +165,7 @@ public function table(Table $table): Table
                         ->action(function ($records) {
                             $deletedRecordsCount = ServiceRequest::query()
                                 ->whereKey($records)
-                                ->when(! auth()->user()->hasRole('authorization.super_admin'), function (Builder $query) {
+                                ->when(! auth()->user()->hasRole(Authenticatable::SUPER_ADMIN_ROLE), function (Builder $query) {
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app-modules/service-management/src/Filament/Resources/ServiceRequestResource/Pages/ListServiceRequests.php]: Added.

File: app-modules/service-management/src/Policies/ServiceRequestPolicy.php [hunk 8]
```diff
@@ -76,7 +76,7 @@ public function view(Authenticatable $authenticatable, ServiceRequest $serviceRe
             return Response::deny('You do not have permission to view this service request.');
         }
 
-        if (! auth()->user()->hasRole('authorization.super_admin')) {
+        if (! auth()->user()->hasRole(Authenticatable::SUPER_ADMIN_ROLE)) {
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app-modules/service-management/src/Policies/ServiceRequestPolicy.php]: Added.

File: app-modules/service-management/src/Policies/ServiceRequestPolicy.php [hunk 9]
```diff
@@ -92,7 +92,7 @@ public function view(Authenticatable $authenticatable, ServiceRequest $serviceRe
 
     public function create(Authenticatable $authenticatable): Response
     {
-        if (! auth()->user()->hasRole('authorization.super_admin')) {
+        if (! auth()->user()->hasRole(Authenticatable::SUPER_ADMIN_ROLE)) {
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app-modules/service-management/src/Policies/ServiceRequestPolicy.php]: Added.

File: app-modules/service-management/src/Policies/ServiceRequestPolicy.php [hunk 10]
```diff
@@ -116,7 +116,7 @@ public function update(Authenticatable $authenticatable, ServiceRequest $service
             return Response::deny('Closed service request cannot be edited.');
         }
 
-        if (! auth()->user()->hasRole('authorization.super_admin')) {
+        if (! auth()->user()->hasRole(Authenticatable::SUPER_ADMIN_ROLE)) {
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app-modules/service-management/src/Policies/ServiceRequestPolicy.php]: Added.

File: app-modules/service-management/src/Policies/ServiceRequestPolicy.php [hunk 11]
```diff
@@ -136,7 +136,7 @@ public function delete(Authenticatable $authenticatable, ServiceRequest $service
             return Response::deny('You do not have permission to delete this service request.');
         }
 
-        if (! auth()->user()->hasRole('authorization.super_admin')) {
+        if (! auth()->user()->hasRole(Authenticatable::SUPER_ADMIN_ROLE)) {
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app-modules/service-management/src/Policies/ServiceRequestPolicy.php]: Added.

File: app-modules/service-management/src/Policies/ServiceRequestPolicy.php [hunk 12]
```diff
@@ -156,7 +156,7 @@ public function restore(Authenticatable $authenticatable, ServiceRequest $servic
             return Response::deny('You do not have permission to restore this service request.');
         }
 
-        if (! auth()->user()->hasRole('authorization.super_admin')) {
+        if (! auth()->user()->hasRole(Authenticatable::SUPER_ADMIN_ROLE)) {
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app-modules/service-management/src/Policies/ServiceRequestPolicy.php]: Added.

File: app-modules/service-management/src/Policies/ServiceRequestPolicy.php [hunk 13]
```diff
@@ -176,7 +176,7 @@ public function forceDelete(Authenticatable $authenticatable, ServiceRequest $se
             return Response::deny('You do not have permission to permanently delete this service request.');
         }
 
-        if (! auth()->user()->hasRole('authorization.super_admin')) {
+        if (! auth()->user()->hasRole(Authenticatable::SUPER_ADMIN_ROLE)) {
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app-modules/service-management/src/Policies/ServiceRequestPolicy.php]: Added.

File: app/Jobs/CreateTenantUser.php [hunk 14]
```diff
@@ -83,7 +84,7 @@ public function handle(): void
                 }
             }
 
-            $user->roles()->sync(Role::where('name', 'authorization.super_admin')->firstOrFail());
+            $user->roles()->sync(Role::where('name', Authenticatable::SUPER_ADMIN_ROLE)->firstOrFail());
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app/Jobs/CreateTenantUser.php]: Added.

File: app/Providers/AuthServiceProvider.php [hunk 15]
```diff
@@ -60,7 +60,7 @@ public function boot(): void
         $this->registerPolicies();
 
         Gate::after(function (Authenticatable $authenticatable, string $ability, bool|null|Response $result, mixed $arguments) {
-            return $authenticatable->hasRole('authorization.super_admin') && ! $result instanceof FeatureAccessResponse
+            return $authenticatable->hasRole(Authenticatable::SUPER_ADMIN_ROLE) && ! $result instanceof FeatureAccessResponse
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app/Providers/AuthServiceProvider.php]: Added.

File: database/seeders/SampleSuperAdminUserSeeder.php [hunk 16]
```diff
@@ -47,7 +48,7 @@ class SampleSuperAdminUserSeeder extends Seeder
     public function run(): void
     {
         if (app()->environment('local')) {
-            $superAdminRole = Role::where('name', 'authorization.super_admin')->firstOrFail();
+            $superAdminRole = Role::where('name', Authenticatable::SUPER_ADMIN_ROLE)->firstOrFail();
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply database/seeders/SampleSuperAdminUserSeeder.php]: Added.

File: app-modules/authorization/tests/Feature/Filament/Resources/User/EditUserTest.php [hunk 17]
```diff
@@ -147,3 +152,40 @@
...
+
+    livewire(RolesRelationManager::class, [
+        'ownerRecord' => $second,
+        'pageClass' => EditUser::class,
+    ])
+        ->mountTableAction(AttachAction::class)
+        ->assertFormFieldExists('recordId', 'mountedTableActionForm', function (Select $select) {
+            $options = $select->getSearchResults(Authenticatable::SUPER_ADMIN_ROLE);
+
+            return empty($options) ? true : false;
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app-modules/authorization/tests/Feature/Filament/Resources/User/EditUserTest.php]: Updated

File: app-modules/authorization/tests/Feature/Filament/Resources/User/EditUserTest.php [hunk 18]
```diff
@@ -147,3 +152,40 @@
...
+        })->assertSuccessful();
```
[the remark this answers is not in the document]
Comment (kandarp-canyon) [reply app-modules/authorization/tests/Feature/Filament/Resources/User/EditUserTest.php]: Removed.

**Pull request merged**